proposals/TypeParameters.md

```diff
@@ -0,0 +1,70 @@
+# Type Parameters Specification
+
+Currently we don't have a way to specify the binding of the type parameters on a function call. They need to be inferred based on the context. 
+
+
+Lets start with a few examples
+
+
+```dataweave
+fun head<T>(a: Array<T>): T = a[0]
+
+var a = head([random() as String])
+```
+
+In this case the type of a is going to be inferred as `String` this is because the type of head function is 
+
+`<T>(Array<T>) -> T` and is being applied with `(Array<String>)`
+
+When this happens in order to infer the type of T a set of constrains are built
+
+
+
+| Expected | Actural | 
+| -------- | --------|
+| T        | String  |
+
+
+Then the solution is simple `T` needs to be of Type `String` then it will substitue the result with this solution. `T <= String` 
+
+This is a very simple example but it shows the basics of how the type parameter inference works.
+
+
+Now lets see a case where these doesn't work as expected.
+
+
+
+```
+fun defaultValue<T>(a: T) : (T) -> T = (value: T) -> if(value == null) a else value      
```

Comment:
This is indeed a comprehensive solution, but I cannot think of a practical application in which a mere function overload would serve the same purpose, without increasing the verbosity of the function call (talking in the context of using DataWeave as a transformation language within the scope of an integration solution. Nevertheless, It does make more sense when looking at DataWeave as a GP language, though...)

Reply:
Function overload doesn't fix the problem because in this case you would need to do it for all the posible values of T that is infinit.  Most probably this won't happen in a transform script but it can happen when people start building libraries with helper functions

Reply:
Agreed